### PR TITLE
Transactional saving of volume annotations

### PR DESCRIPTION
**Pull request by philippotto**
### Description:
- Introduces sensible transaction points for volume annotations (--> also means less versions which will be helpful in the version restore view :tada:). This is done by waiting for small time windows (1 second) in which no volume data was mutated **and** no dirty volume buckets need to be downloaded. This is similar to how the PushQueue operated, anyway, but it chunked buckets into 32 batches later which destroyed any transactional semantics. Reusing the state machine of the undo mechanism would be nicer, since it is strictly user-interaction based. However, the implementation of that would be far more complicated.
- Improved speed of bucket compression by using larger payloads for the webworkers (an isolated benchmark with reasonable inputs showed that the bucket compression is now 5 to 10 times faster).
- Removes AsyncTaskQueue as its sequentiality and failure-mechanism aren't needed anymore. [This abstraction was originally introduced to deal with failing network requests.](https://github.com/scalableminds/webknossos/blob/65319c4de0bb0193af4a47fc1ec746435395e9f9/app/assets/javascripts/oxalis/model/binary/layers/wk_layer.js#L96) However, that part is not handled by the AsyncTaskQueue anymore.
- The error handling in the PushQueue was hardened and unified with the error handling in the sagas (by escalating potential exceptions to the sagas).
- Fixes a small bug in the save-saga which could lead to an accumulating backlog of save queue items that would grow faster than they would be saved. This could happen when brushing a high volume of data and only relying on the auto-save mechanism. That mechanism did only send 1 request every 30 seconds (even when multiple requests would be necessary to completely save the state of that time).
- The save button now also warns when the PushQueue is starving (i.e., constantly unable to bundle up a transaction, because a bucket was edited again). Previously, this was achieved by providing a max_time parameter to _.debounce. However, since we want sensible transactions, this is not an option anymore.
- Move some modules from libs to libs/async.
- Lower saving batch constants for volume tracings, because the backend needs more time to process the update actions (see discussion [here](https://scm.slack.com/archives/C5AKLAV0B/p1692177165085429))
- Improve progress indicator during saving (the tooltip for the save button will also show when buckets need to be downloaded during saving, because this can take a noticable amount of time).

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- do a volume annotation (with and without fallback)
- set the network speed to very slow
- reload the tab during save (large volume changes are necessary to cause multiple requests)
- ensure that the volume data looks consistent (zoom in and out to check all mags)
- additionally/alternatively:
  - brush a large area
  - save
  - look at the version restore view. there should be one version (maybe other versions for the new segment id and for updateTracing, but not more)
  - compare the above to master.webknossos.xyz where lots of versions will be created. previewing one version and zooming to different mags will show hard-cuts in the brushed data.

### TODOs:
- [X] test volume speed on dev instance (which will use prod mode) to ensure that it didn't regress
- [X] verify that new save requests don't get too large. since updateBucket actions are way larger than skeleton update actions, some tuning might be necessary here. <-- no problem, since the compression does a good job of keeping the size small

### Issues:
- hopefully fixes [#7256](https://github.com/scalableminds/webknossos/issues/7256)
- contributes to #6495
- to close the issue fully, the version history view should show the transactions (the back-end needs to store these first, though)

------
(Please delete unneeded items, merge only when none are left open)
- [ ] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

### CODE REVIEW

File: frontend/javascripts/oxalis/model/sagas/save_saga_constants.ts [hunk 1]
```diff
@@ -11,5 +11,5 @@ export const UNDO_HISTORY_SIZE = 20;
 export const SETTINGS_RETRY_DELAY = 15 * 1000;
 export const SETTINGS_MAX_RETRY_COUNT = 20; // 20 * 15s == 5m
 
-export const maximumActionCountPerBatch = 5000;
-export const maximumActionCountPerSave = 15000;
+export const maximumActionCountPerBatch = 20;
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/sagas/save_saga_constants.ts:14]: todo: revert

File: frontend/javascripts/oxalis/model/sagas/save_saga_constants.ts [hunk 2]
```diff
@@ -11,5 +11,5 @@ export const UNDO_HISTORY_SIZE = 20;
 export const SETTINGS_RETRY_DELAY = 15 * 1000;
 export const SETTINGS_MAX_RETRY_COUNT = 20; // 20 * 15s == 5m
 
-export const maximumActionCountPerBatch = 5000;
-export const maximumActionCountPerSave = 15000;
+export const maximumActionCountPerBatch = 5000; //20; // 5000
+export const maximumActionCountPerSave = 15000; // 60; // 15000
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/sagas/save_saga_constants.ts:15]:
```suggestion
export const maximumActionCountPerBatch = 5000;
export const maximumActionCountPerSave = 15000;
```

File: frontend/javascripts/oxalis/model/sagas/save_saga.ts [hunk 3]
```diff
@@ -151,9 +159,19 @@ function getRetryWaitTime(retryCount: number) {
...
+): Saga<number> {
+  /*
+   * Saves a reasonably-sized part of the save queue (that corresponds to the
+   * tracingId) to the server (plus retry-mechanism).
+   * The saga returns the number of save queue items that were saved.
+   */
+
   const fullSaveQueue = yield* select((state) => selectQueue(state, saveQueueType, tracingId));
   const saveQueue = sliceAppropriateBatchCount(fullSaveQueue);
+  console.log("saving", saveQueue.length, "items out of", fullSaveQueue.length);
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/sagas/save_saga.ts:174]:
```suggestion
```

File: frontend/javascripts/oxalis/model/bucket_data_handling/data_cube.ts [hunk 4]
```diff
@@ -679,6 +679,7 @@ class DataCube {
   }
 
   triggerPushQueue() {
+    console.log("triggerPushQueue");
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/bucket_data_handling/data_cube.ts:682]:
```suggestion
```

File: frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts [hunk 5]
```diff
@@ -77,42 +97,75 @@ class PushQueue {
...
 
-  pushImpl = async () => {
-    await this.cube.temporalBucketManager.getAllLoadedPromise();
-
-    if (!this.sendData) {
-      return;
+  pushImpl = function* (this: PushQueue) {
+    try {
+      console.log("throw in pushimpl");
+      console.log("pushImpl start");
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts:103]:
```suggestion
```

File: frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts [hunk 6]
```diff
@@ -77,42 +97,75 @@ class PushQueue {
...
+      this.flushAndSnapshot();
+    } catch (error) {
+      // The above code is critical for saving volume data. Because the
+      // code is invoked asynchronously, there won't be a default error
+      // handling in case it crashes due to a bug.
+      // Therefore, escalate the error manually so that the sagas will crash
+      // (notifying the user and stopping further potentially undefined behavior).
+      Store.dispatch(escalateErrorAction(error));
     }
+    console.log("pushImpl end");
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts:120]:
```suggestion
```

File: frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts [hunk 7]
```diff
@@ -77,42 +97,75 @@ class PushQueue {
...
 
-    while (this.pendingQueue.size) {
-      let batchSize = Math.min(COMPRESSING_BATCH_SIZE, this.pendingQueue.size);
-      const batch: DataBucket[] = [];
+  private flushAndSnapshot() {
+    this.waitTimeStartTimeStamp = null;
+    // Flush pendingQueue. Note that it's important to do this synchronously.
+    // If other actors could add to queue concurrently, the front-end could
+    // send an inconsistent state for a transaction.
+    console.log("Flush pending queue with size:", this.pendingQueue.size);
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts:128]:
```suggestion
```

File: frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts [hunk 8]
```diff
@@ -1,74 +1,91 @@
...
-  constructor(cube: DataCube, sendData: boolean = true) {
+
+  // The pendingQueue contains all buckets that should be:
+  // - snapshotted,
+  // - put into one transaction and then
+  // - saved
+  // That queue is flushed in a debounced manner so that the time of the
+  // snapshot should be suitable for a transaction (since neither WK nor the
+  // user edited the buckets in a certain time window).
+  private pendingQueue: Set<DataBucket>;
```
Comment (normanrz) [inline frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts:29]: Why is the queue a `Set`?

**Review by normanrz — COMMENTED** [review]
This code is a piece of art! Thanks for writing all the descriptive comments. I think I understand what is roughly going on and look forward to the walkthrough.

File: frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts (hunk 8 above)
Comment (philippotto) [reply frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts]: The order of the buckets doesn't matter within a transaction, which is why this is a set. However, the mismatch is obviously still an eye sore. I renamed it to `pendingBuckets` now. The class `PushQueue` might also be ripe for a new name, but then the `PullQueue` loses its "name-sibling" :sweat_smile: Also, I'm not sure how to call it. `BucketTransactionBundler`? Might be okay to keep the PushQueue as is.

**Review by normanrz — APPROVED** [review]

**Pull request merged**